storage/storage.go

```diff
@@ -35,3 +52,14 @@ func (f Fanout) Append(s *model.Sample) {
 		a.Append(s)
 	}
 }
+
+// Throttled returns true if at least one of the SampleAppenders in the Fanout
+// slice is throttled.
+func (f Fanout) Throttled() bool {
```

Comment:
Well, if one storage can take it, I want to insert and consequentially do the work. I use throttle to determine whether the result of my work can be used. Which it can by one out of several storages. If none of them does, than I don't need to do the work.

> It requires to copy all samples (up to 500k/s) into a vector before handing it over to the storage, with no way of reusing it.

No. This is per scrape per target. Thus those vectors will be about ~1k samples on average and 10-20k in the extreme cases, which is not really an issue memory-wise. The vector of Sample pointers is perfectly reusable.

> It suggests atomicity that the current storage is not designed to deliver. (How wold you return the information that for three out of the 1000 samples in the vector, there was an out-of-order timestamp?)

A multi-error.
Guarantees can be documented.

> It does not help with the idea of first finding out if the storage is ready to ingest and then do the work of a potentially very large scrape.

Suppose N storages in a fanout. If one or more can take samples, do the scrape and call `Append(<all samples of the scrape>)`. Fanout calls `Append()` for all M storages that are not throttled.